util/dvsim/CircularList.py

```diff
@@ -0,0 +1,123 @@
+
+    This class is implemented as a circular linked list. Each item in the list
+    is an object of the Node, including the head pointer, which has null item.
+    As new items are added at the end, the linked list is constructed with the
+    help of the tail pointer. The last item is linked back to the head pointer.
+    Any item in the list can be removed. The class is iterable, but it will
+    only loop through the list of items once. To indefinitely loop through the
+    list of items, circling back to the start, use the get_next_item() method.
+    '''
+    def __init__(self):
+        '''Initialize the head, tail and circular iteration pointers.'''
+        # 'head' is always an empty node.
+        self.head = Node()
+        self.head.next_node = self.head
+
+        # 'tail' is the last item in the list
+        self.tail = self.head
+
+        # 'ptr' is the current iterated item
+        self.ptr = self.head
+
+        # Maintain a count of items in the list.
+        self.count = 0
+
+    def __str__(self):
+        items = []
+        current = self.head.next_node
+        while current != self.head:
+            items.append(current.item.__str__())
+            current = current.next_node
+        return items.__str__()
+
+    def __iter__(self):
+        '''Iterate over the list, stopping at the last entry.'''
+        return CircularListIterator(self)
+
+    def __len__(self):
+        '''Returns the number of items in the list.'''
+        return self.count
+
+    def __bool__(self):
+        '''Returns True if there are items in the list, else False.'''
+        return bool(self.count)
+
+    def add(self, item):
+        '''Adds a new item at the end of the list.'''
+        node = Node(item=item, next_node=self.head)
+        self.tail.next_node = node
+        self.tail = node
+        self.count += 1
+
+    append = add
+    '''Alias of add().'''
+
+    def extend(self, items):
+        '''Extends the list with multiple items.'''
+        for item in items:
+            self.append(item)
+
+    def remove(self, item):
+        '''Removes the item from the list.'''
+        previous = self.head
+        current = self.head.next_node
+        while current != self.head:
+            if current.item == item:
+                previous.next_node = current.next_node
+                if self.tail == current:
+                    self.tail = previous
+                if self.ptr == current:
+                    self.ptr = previous
+                del current
+                self.count -= 1
+                return
+            previous = current
+            current = current.next_node
+        raise ValueError("List does not contain item {}".format(item))
+
+    def get_next_item(self):
+        '''Circularly iterate over the list and return an item.
+
+        This iteration loops back forever, returning an item if available.
+        Returns None if the list is empty.
```

Comment:
Maybe raise an exception like `StopIteration` if the list is empty? Then we can store an `Optional[X]` in the circular list without any gymnastics.